File: module1/module 1 c++/poo_basic_class.cpp
#include <iostream>
#include<cstdio>
#include <string>
#include <vector>

using namespace std;
typedef unsigned char ui8;
typedef unsigned int ui32;

class Gato
{
public:
    string nombre;
    int edad;
    int peso;
    string color;

    Gato(string nom, int ed, int pes, string col)
    {
        nombre = nom;
        edad  = ed;
        peso  = pes;
        color = col;
    }

    char* detalles()
    {
        char* buffer = new char[100];
        sprintf(buffer, "%s es un gato de %d anios color %s que pesa %d kilos\n",nombre.c_str(), edad,color.c_str(), peso);
        return buffer;
    }
};


int main()
{
    Gato tom("Tom", 3, 7, "Cafe");

    char* arr = tom.detalles();
    cout << arr << endl;
    return 0;
}

File: module1/poo_basic_class.py
# Programación orientada a objetos > Conceptos básicos de POO y UML

class Gato:
    ''' Crea un objeto del tipo gato con características básicas '''
    # Recibimos la instancia como primer argumento
    # Podemos nombrar al argumento diferente pero por convención usamos self
    def __init__(self, nombre:str, edad:int, peso:float, color:str):
        # VARIABLES DE INSTANCIA
        self.nombre:str = nombre
        self.edad:int = edad
        self.peso:float = peso
        self.color:str = color
    
    # El argumento self se pasa de manera automática
    def detalles(self) -> str:
        ''' Muestra los detalles del objeto tipo gato '''
        # A continuación se utiliza el método format para reemplazar valores de
        # variables en un string
        return '{} es un gato de {} años color {} que pesa {} kilos.'.format(
            self.nombre,
            self.edad,
            self.color,
            self.peso
        )

# Creamos el objeto del tipo gato
tom = Gato('Tom', 3, 7, 'café')
print(tom.detalles())

# Incluso podemos acceder al método detalles y pasar la INSTANCIA 'tom'
print(Gato.detalles(tom))